### PR TITLE
[P8-45/Errors] Add Errors to symbolic analysis phase of LCGS

### PR DESCRIPTION
**Pull request by NicEastvillage**
The LCGS compiler will now produce errors during the symbolic analysis phase. That includes
* Unknown identifiers
* Misuse of identifiers
* Misuse of relabeling

The parsing still doesn't produce errors and the errors does not include line number (yet). This requires rewrite of the parsing.

### CODE REVIEW

File: src/lcgs/ir/symbol_checker.rs [hunk 1]
```diff
@@ -47,19 +52,19 @@ impl<'a> SymbolChecker<'a> {
...
+    /// Checks and evaluates an expressions. This is only used in [CheckMode::Const]
     /// where we assume the expression can be reduced to a value already during symbol checking.
-    pub fn check_eval(&self, expr: &Expr) -> Result<i32, ()> {
+    pub fn check_eval(&self, expr: &Expr) -> Result<i32, SymbolError> {
         let checked = self.check(expr)?;
         if let ExprKind::Number(n) = checked.kind {
             Ok(n)
         } else {
-            Err(())
+            panic!("Constant expression was not reduced to a single number.")
```
Comment (jenrik) [inline src/lcgs/ir/symbol_checker.rs:62]: Shouldn't this be an error?

**Review by jenrik — APPROVED** [review]
Instead of printing the Error trait, it might be more idiomatic to implement the Display trait on the Error type and print that instead.

The core functionality of the PR looks good

**Comment by NicEastvillage** [comment]
> it might be more idiomatic to implement the Display trait on the Error type and print that instead.

I did. `src/lcgs/ir/error.rs` line 12. What do you mean?

File: src/lcgs/ir/symbol_checker.rs (hunk 1 above)
Comment (NicEastvillage) [reply src/lcgs/ir/symbol_checker.rs:62]: Nope. The previous phase should have caught this error already. If it is not a single number here, something is wrong within our compiler.

**Comment by jenrik** [comment]
@NicEastvillage I don't know how I missed that, well then that's all good then

**Review by cogitantium — APPROVED** [review]

**Review by jenrik — APPROVED** [review]

**Pull request merged**